configs/config.go

```diff
@@ -26,53 +34,116 @@ type ConnInfo struct {
+)
+
 // Load configs from env
-func LoadFromEnv() *Configs {
+func LoadFromEnv() {
 	err := godotenv.Load()
 	if err != nil {
 		log.Fatal("Error loading .env file")
 	}
 
-	return &Configs{
-		Env:  os.Getenv(constant.ENV),
-		Port: os.Getenv(constant.PORT),
-		DB: &ConnInfo{
-			Host: os.Getenv(constant.DB_HOST),
-			Port: os.Getenv(constant.DB_PORT),
-			User: os.Getenv(constant.DB_USER),
-			Pass: os.Getenv(constant.DB_PASS),
-			Name: os.Getenv(constant.DB_NAME),
+	Config = Configs{
+		Env:  envDefaultValueString("development", os.Getenv(constant.ENV)),
+		Port: envDefaultValueString("8080", os.Getenv(constant.PORT)),
+		DB: ConnInfo{
+			Host:        envDefaultValueString("localhost", os.Getenv(constant.DB_HOST)),
+			Port:        envDefaultValueString("5432", os.Getenv(constant.DB_PORT)),
+			User:        envDefaultValueString("postgres", os.Getenv(constant.DB_USER)),
+			Pass:        envDefaultValueString("secret", os.Getenv(constant.DB_PASS)),
+			Name:        envDefaultValueString("postgres", os.Getenv(constant.DB_NAME)),
+			MaxOpenConn: envDefaultValueInt(100, os.Getenv(constant.DB_MAX_OPEN_CONN)),
+			MaxIdleConn: envDefaultValueInt(5, os.Getenv(constant.DB_MAX_IDLE_CONN)),
+			MaxLifeTime: envDefaultValueInt(15, os.Getenv(constant.DB_MAX_LIFE_TIME)),
 		},
-		Redis: &ConnInfo{
-			Host: os.Getenv(constant.REDIS_HOST),
-			Port: os.Getenv(constant.REDIS_PORT),
-			User: os.Getenv(constant.REDIS_USER),
-			Pass: os.Getenv(constant.REDIS_PASS),
+		Redis: ConnInfo{
+			Host: envDefaultValueString("localhost", os.Getenv(constant.REDIS_HOST)),
+			Port: envDefaultValueString("6379", os.Getenv(constant.REDIS_PORT)),
+			User: envDefaultValueString("admin", os.Getenv(constant.REDIS_USER)),
+			Pass: envDefaultValueString("secret", os.Getenv(constant.REDIS_PASS)),
 		},
-		JWT: &JWT{
-			SecretKey: os.Getenv(constant.JWT_SECRET_KEY),
+		JWT: JWT{
+			SecretKey: envDefaultValueString("secretJwtKey", os.Getenv(constant.JWT_SECRET_KEY)),
 		},
 	}
 }
 
-func LoadFromVault() *Configs {
-	return &Configs{}
+func LoadFromVault(address, token string) {
+	config := vault.DefaultConfig()
+	config.Address = address
+
+	client, err := vault.NewClient(config)
+	if err != nil {
+		log.Fatalf("unable to initialize Vault client: %v", err)
+	}
+
+	client.SetToken(token)
+
+	ctx := context.Background()
+
+	secret, err := client.KVv2(os.Getenv("SERVICE_NAME")).Get(ctx, os.Getenv("SECRET_NAME"))
+	if err != nil {
+		log.Fatalf("unable to read secret: %v", err)
+	}
+
+	b, err := sonic.Marshal(secret.Data)
+	if err != nil {
+		log.Fatalf("unable to read secret: %v", err)
+	}
+
+	var temp Configs
+	err = sonic.Unmarshal(b, &temp)
+	if err != nil {
+		log.Fatalf("unable to read secret: %v", err)
+	}
+
+	// remove mutex if you don't load the config anymore
+	Mutex.Lock()
+	Config = temp
+	Mutex.Unlock()
+}
+
+func (conf Configs) IsDevelopment() bool {
+	return strings.ToLower(conf.Env) == "development"
 }
 
-func (conf *Configs) IsDevelopment() bool {
```

Comment:
Keep it pointer, maybe in the future we need to change the config in somewhere so every action that call this config will get new config

Reply:
In this case we don't need pointer, like [stdlib time](https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/time/time.go;l=250) case.

And because Config is public, we can edit it directly.

Reply:
Alrigh let's keep this for now